### PR TITLE
Do not eat ErrorReport messages related to repairs

### PR DESCRIPTION
**Pull request by jmpesp**
If an ErrorReport is received for a live repair, this should cause the Upstairs to kick out the associated Downstairs, and attempt to reconnect. The Downstairs that sends such an ErrorReport will abort its tasks and wait for the Upstairs to reconnect, so if the Upstairs does not kick out the associated Downstairs, then it will get stuck in a repair task waiting for the repair ack response, and no IO can flow due to how repair related dependencies work.

### CODE REVIEW

**Comment by jmpesp** [comment]
I believe this fixes https://github.com/oxidecomputer/crucible/issues/837 but I haven't tested it yet

File: upstairs/src/dummy_downstairs_tests.rs [hunk 1]
```diff
@@ -1880,4 +1881,710 @@ pub(crate) mod protocol_test {
...
+                        .read(Block::new_512(0), buffer)
+                        .await
+                        .unwrap();
+                });
+            }
+
+            if i < MAX_ACTIVE_COUNT {
+                // Before flow control kicks in, assert we're seeing the read
+                // requests
+                bail_assert!(matches!(
```
Comment (leftwo) [inline upstairs/src/dummy_downstairs_tests.rs:1925]: Do we need to ignore ping messages too?

File: upstairs/src/dummy_downstairs_tests.rs [hunk 2]
```diff
@@ -1880,4 +1881,710 @@ pub(crate) mod protocol_test {
...
+                    ds1_messages.recv().await.unwrap(),
+                    Message::ReadRequest { .. },
+                ));
+            } else {
+                // After flow control kicks in, we shouldn't see any more
+                // messages
+                match ds1_messages.try_recv() {
+                    Err(TryRecvError::Empty) => {}
+                    Err(TryRecvError::Disconnected) => {}
+                    x => {
```
Comment (leftwo) [inline upstairs/src/dummy_downstairs_tests.rs:1935]: Do we need to ignore ping messages here too?

For all these cases really, for ds2 and ds3, they could all receive a ping, right?

File: upstairs/src/lib.rs [hunk 3]
```diff
@@ -498,18 +498,53 @@ async fn process_message(
...
+            session_id,
             job_id,
             error,
         } => {
-            // XXX currently, this error report goes nowhere except to the log.
-            // The Upstairs should track this for each Downstairs.
+            // The Upstairs should not consider a job completed until it has
+            // returned an Ok result, and should therefore log and eat all
+            // ErrorReport messages here. This will change in the future when
+            // the Upstairs tracks the number of errors per Downstairs, and acts
```
Comment (leftwo) [inline upstairs/src/lib.rs:509]: Let's file an issue for this lest we forget it.

File: upstairs/src/lib.rs [hunk 4]
```diff
@@ -498,18 +498,53 @@ async fn process_message(
...
+            // However, there is one case (see `check_message_for_abort` in
+            // downstairs/src/lib.rs) where the Upstairs **does** need to
+            // act: when a repair job in the Downstairs fails, that Downstairs
+            // aborts itself and reconnects.
+            let _active = u.active.lock().await;
+            let ds = u.downstairs.lock().await;
+
+            if let Some(job) = ds.ds_active.get(job_id) {
+                if job.work.is_repair() {
+                    // Return the is error and let the previously written error
```
Comment (leftwo) [inline upstairs/src/lib.rs:525]: Extra `is` here

File: upstairs/src/lib.rs [hunk 5]
```diff
@@ -498,18 +498,53 @@ async fn process_message(
...
+                    // processing code work.
+                    cdt::ds__repair__done!(|| (job_id, client_id as u64));
+
+                    // XXX uncomment this to see the upstairs disconnect this
```
Comment (leftwo) [inline upstairs/src/lib.rs:529]: We should take out the XXX comments here, I think once we know this does what we expect
we don't need to comments.  I don't know why (other than testing) we would want to get the
upstairs stuck in live repair.

File: upstairs/src/lib.rs [hunk 6]
```diff
@@ -4193,6 +4228,8 @@ impl Downstairs {
                             self.downstairs_errors
                                 .insert(client_id, errors + 1);
                         }
+
+                        // If an repair job errors, mark that downstairs as bad
```
Comment (leftwo) [inline upstairs/src/lib.rs:4232]: Nit: If `a` repair job fails..

File: upstairs/src/lib.rs [hunk 7]
```diff
@@ -4231,6 +4268,7 @@ impl Downstairs {
                             self.downstairs_errors
                                 .insert(client_id, errors + 1);
                         }
+
```
Comment (leftwo) [inline upstairs/src/lib.rs:4271]: Nit: looks like an extra blank line slipped in here..

**Review by leftwo — APPROVED** [review]
I think this is good, it sends the IO back to the repair task directly which means
we don't need a whole new set of plumbing for notifying the repair task that
the task should exit.

I _think_ that's the right thing here... I think...

File: upstairs/src/lib.rs (hunk 5 above)
Comment (jmpesp) [reply upstairs/src/lib.rs:529]: I'm allowed a few XXX, this PR's still in draft! :P

File: upstairs/src/dummy_downstairs_tests.rs (hunk 1 above)
Comment (jmpesp) [reply upstairs/src/dummy_downstairs_tests.rs:1925]: Luckily the task in `spawn_message_receiver` eats pings (and responds to them!) for us.

File: upstairs/src/dummy_downstairs_tests.rs (hunk 2 above)
Comment (jmpesp) [reply upstairs/src/dummy_downstairs_tests.rs:1935]: See other comment :)

File: upstairs/src/lib.rs (hunk 3 above)
Comment (jmpesp) [reply upstairs/src/lib.rs:509]: I filed https://github.com/oxidecomputer/crucible/issues/857, which is a general version of the work required (as the previous error handling code did count the number of errors per Downstairs!).

File: upstairs/src/lib.rs (hunk 4 above)
Comment (jmpesp) [reply upstairs/src/lib.rs]: fixed in 20def81

File: upstairs/src/lib.rs (hunk 6 above)
Comment (jmpesp) [reply upstairs/src/lib.rs]: fixed in 5f10b2c

File: upstairs/src/lib.rs (hunk 7 above)
Comment (jmpesp) [reply upstairs/src/lib.rs:4271]: That was an intention separation, but I added a comment in 940cacb

**Pull request merged**